{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5349bc9a-1946-4809-bcf6-a29afed522e7",
   "metadata": {},
   "source": [
    "# Call, annotate, and analyze peaks for each cell type\n",
    "\n",
    "To analyze the scATAC-seq data, we'll need to call peaks and identify TF motifs in each peak. To facilitate downstream analyses, a single set of peaks will be called for all cells obtained for each cell type. This will give us a single set of features to use for differential testing for each type.\n",
    "\n",
    "Once we have peaks, we'll analyze them from 3 angles: \n",
    "1. Identify differentially accessible sites induced by drug treatment at each time point.\n",
    "2. Calculate enrichment of motif annotations in differentially-acessible peaks.\n",
    "3. Compute Peak-to-gene correlations using the paired scRNA-seq data from our TEA-seq experiments.\n",
    "\n",
    "Along the way, we'll save each of the stages of analysis for downstream work and visualization:\n",
    "- Peaks and annotations\n",
    "- Differential peaks and annotation enrichments\n",
    "- Peak-to-gene correlations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6a29243-663c-42cc-ae9e-bd6db6f32258",
   "metadata": {},
   "source": [
    "# Setup\n",
    "\n",
    "To call peaks, we'll need to install MACS2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5242ee7e-2833-4e69-9abf-d63a00e300f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "system(\"pip install --upgrade --force-reinstall MACS2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c54a7bb7-1940-446e-a43b-d1327d09856b",
   "metadata": {},
   "source": [
    "We'll also need the JASPAR2020 package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b40113ba-bb28-413d-9222-39170af1e70f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "Bioconductor version 3.17 (BiocManager 1.30.22), R 4.3.1 (2023-06-16)\n",
      "\n",
      "Installing package(s) 'JASPAR2020'\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "BiocManager::install(\"JASPAR2020\", update = FALSE, ask = FALSE, force = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "778c921c-112c-4c95-a734-8f886b501b47",
   "metadata": {},
   "source": [
    "## Load packages\n",
    "\n",
    "hise: The Human Immune System Explorer R SDK package  \n",
    "dplyr: Dataframe handling functions   \n",
    "ArchR: scATAC-seq analysis  \n",
    "purrr: Functional programming tools  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d4b0c889-cc81-4412-8ebb-97545506db86",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "quiet_library <- function(...) { suppressPackageStartupMessages(library(...)) }\n",
    "quiet_library(hise)\n",
    "quiet_library(dplyr)\n",
    "quiet_library(ArchR)\n",
    "quiet_library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "quiet_library(purrr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73e4a3e7-cdd3-475e-a54a-cca61151e662",
   "metadata": {},
   "source": [
    "## Retrieve files\n",
    "\n",
    "Now, we'll use the HISE SDK package to retrieve the TEA-seq ArchR Projects based on their file UUIDs. These will be placed in the `cache/` subdirectory by default."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f69e39ef-a0d8-4d09-9983-412c01fa3142",
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_file_uuids <- list(\n",
    "    \"11235534-d09d-4c57-a648-20fb13317eab\",\n",
    "    \"2d1a00ca-f1f6-41c1-9691-f37916fad00c\",\n",
    "    \"365045be-e8a6-4a4d-9fe1-b31b7593799a\",\n",
    "    \"403e1064-34ea-4992-8752-6d1ddb9fb614\",\n",
    "    \"49d66578-bc16-4840-871c-25de96456f83\",\n",
    "    \"f1a32e62-d2e9-4052-b971-dd960d605d70\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4baa66f5-cfaa-40f8-bb35-864dcd041832",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Initiating file download for vrdtea_ArchR-t_cd4_cm_2023-10-02.tar\"\n",
      "[1] \"Download successful.\"\n",
      "[1] \"Initiating file download for vrdtea_ArchR-t_cd4_em_2023-10-02.tar\"\n",
      "[1] \"Download successful.\"\n",
      "[1] \"Initiating file download for vrdtea_ArchR-t_cd4_treg_2023-10-02.tar\"\n",
      "[1] \"Download successful.\"\n",
      "[1] \"Initiating file download for vrdtea_ArchR-t_cd4_naive_2023-10-02.tar\"\n",
      "[1] \"Download successful.\"\n",
      "[1] \"Initiating file download for vrdtea_ArchR-t_cd8_naive_2023-10-02.tar\"\n",
      "[1] \"Download successful.\"\n",
      "[1] \"Initiating file download for vrdtea_ArchR-t_cd8_memory_2023-10-02.tar\"\n",
      "[1] \"Download successful.\"\n"
     ]
    }
   ],
   "source": [
    "fres <- hise::cacheFiles(\n",
    "    atac_file_uuids\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "965a99e0-62b0-40cb-800d-67711c9d0458",
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_tar_files <- map(fres, \"filePath\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fb395cf3-8cb3-4ebe-adcc-4bd2044cd2fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "walk(\n",
    "    atac_tar_files,\n",
    "    function(tf) {\n",
    "        command <- paste(\"tar -xf\", tf)\n",
    "        system(command)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13308355-76ec-4c05-b824-862654b11a02",
   "metadata": {},
   "source": [
    "Note: un-tar-ing the files moves them to the `output` directory, based on their original filenames."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "50c0ca2f-5bee-444b-aa27-afe03dff4bb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "type_paths <- list.files(\n",
    "    \"output\",\n",
    "    full.names = TRUE\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b672b16f-98b2-424d-b478-8530df307fb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'t_cd4_cm'</li><li>'t_cd4_em'</li><li>'t_cd4_naive'</li><li>'t_cd4_treg'</li><li>'t_cd8_memory'</li><li>'t_cd8_naive'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 't\\_cd4\\_cm'\n",
       "\\item 't\\_cd4\\_em'\n",
       "\\item 't\\_cd4\\_naive'\n",
       "\\item 't\\_cd4\\_treg'\n",
       "\\item 't\\_cd8\\_memory'\n",
       "\\item 't\\_cd8\\_naive'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 't_cd4_cm'\n",
       "2. 't_cd4_em'\n",
       "3. 't_cd4_naive'\n",
       "4. 't_cd4_treg'\n",
       "5. 't_cd8_memory'\n",
       "6. 't_cd8_naive'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"t_cd4_cm\"     \"t_cd4_em\"     \"t_cd4_naive\"  \"t_cd4_treg\"   \"t_cd8_memory\"\n",
       "[6] \"t_cd8_naive\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cell_types <- sub(\".+-(.+)_20.+\", \"\\\\1\", type_paths)\n",
    "cell_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "899aed64-302f-4d91-81b4-ed9c4254c4d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Successfully loaded ArchRProject!\n",
      "\n",
      "Successfully loaded ArchRProject!\n",
      "\n",
      "Successfully loaded ArchRProject!\n",
      "\n",
      "Successfully loaded ArchRProject!\n",
      "\n",
      "Successfully loaded ArchRProject!\n",
      "\n",
      "Successfully loaded ArchRProject!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "type_proj <- map(\n",
    "    type_paths,\n",
    "    loadArchRProject,\n",
    "    showLogo = FALSE\n",
    ")\n",
    "names(type_proj) <- cell_types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a747d7c-b327-47de-bf62-dc1e14a9edf3",
   "metadata": {},
   "source": [
    "# Run analysis per cell type"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9d5638a-a233-48bc-85b0-f749733b0ec3",
   "metadata": {},
   "source": [
    "## Peak calling and annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "63a97490-bd3c-40db-a871-afdd7e14d385",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting addArchRVerbose = FALSE\n",
      "\n",
      "Setting default number of Parallel threads to 14.\n",
      "\n",
      "Setting default genome to Hg38.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "addArchRVerbose(verbose = FALSE)\n",
    "addArchRThreads(14)\n",
    "addArchRGenome(\"hg38\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "329a4bc9-6bc5-4da9-bd45-cfe16a15d036",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table style=\"width: 100%;\"><tr><td>getMatrixSet {TFBSTools}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2 id='getMatrixSet'>Advanced JASPAR database search functions <code>get_MatrixSet</code></h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>This function fetches matrix data for all matrices in the database \n",
       "matching criteria defined by the named arguments \n",
       "and returns a PFMatrixList object\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre><code class='language-R'>  ## S4 method for signature 'character'\n",
       "getMatrixSet(x, opts)\n",
       "  ## S4 method for signature 'SQLiteConnection'\n",
       "getMatrixSet(x, opts)\n",
       "  ## S4 method for signature 'JASPAR2014'\n",
       "getMatrixSet(x, opts)\n",
       "</code></pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table>\n",
       "<tr><td><code id=\"getMatrixSet_:_x\">x</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a character vector of length 1 for the path of JASPAR SQLite file, \n",
       "a <code>SQLiteConnection</code> object, or a <code>JASPAR2014</code> object.\n",
       "</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"getMatrixSet_:_opts\">opts</code></td>\n",
       "<td>\n",
       "\n",
       "<p>a search options list. See more details below.\n",
       "</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>The search options include three categories:\n",
       "</p>\n",
       "<p>(1) Database basic criterias: \n",
       "</p>\n",
       "<p><code>all=c(TRUE, FALSE)</code> \n",
       "</p>\n",
       "<p><code>ID</code>: a unique identifier for each model. \n",
       "CORE matrices always have a &quot;MAnnnnIDs.Version&quot;.\n",
       "</p>\n",
       "<p><code>name</code>: The name of the transcription factor. \n",
       "As far as possible, the name is based on the standardized \n",
       "Entrez gene symbols. \n",
       "In the case the model describes a transcription factor hetero-dimer, \n",
       "two names are concatenated, such as RXR-VDR. \n",
       "In a few cases, different splice forms of the same gene \n",
       "have different binding specificity: in this case \n",
       "the splice form information is added to the name, \n",
       "based on the relevant literature.\n",
       "</p>\n",
       "<p><code>collection=c(\"CORE\", \"CNE\", \"PHYLOFACTS\", \"SPLICE\", \n",
       "    \"POLII\", \"FAM\", \"PBM\", \"PBM_HOMEO\", \"PBM_HLH\", \"UNVALIDATED\"</code>\n",
       "</p>\n",
       "<p><code>all_versions=c(FALSE,TRUE)</code>: We constantly update the profiles in\n",
       "JASPAR. Some profiles may have multiple versions. \n",
       "By default, only the latest version will be returned.\n",
       "</p>\n",
       "<p><code>species</code>: The species source for the sequences, in Latin (Homo sapiens)\n",
       "or NCBI tax IDs (9606).\n",
       "</p>\n",
       "<p><code>matrixtype=c(\"PFM\", \"PWM\", \"ICM\")</code>\n",
       "</p>\n",
       "<p>(2) Tags based criterias: \n",
       "</p>\n",
       "<p><code>class</code>: Structural class of the transcription factor, \n",
       "based on the TFCaT system. \n",
       "Examples: &quot;Zipper-Type&quot;&quot;, &quot;Helix-Turn-Helix&quot;, etc.\n",
       "</p>\n",
       "<p><code>type</code>: Methodology used for matrix construction: \n",
       "&quot;SELEX&quot;, &quot;ChIP-seq&quot;, &quot;PBM&quot;, etc.\n",
       "</p>\n",
       "<p><code>tax_group</code>: Group of species, \n",
       "currently consisting of &quot;plants&quot;, &quot;vertebrates&quot;, &quot;insects&quot;, \n",
       "&quot;urochordat&quot;, &quot;nematodes&quot;, &quot;fungi&quot;.\n",
       "</p>\n",
       "<p><code>family</code>: Structural sub-class of the transcription factor, \n",
       "based on the TFCaT system.\n",
       "</p>\n",
       "<p><code>Acc</code>: A representative protein accession number in Genbank \n",
       "for the transcription factor. \n",
       "Human takes precedence if several exists. \n",
       "</p>\n",
       "<p><code>medline</code>: relevant publication reporting the sites \n",
       "used in the mode building.\n",
       "</p>\n",
       "<p><code>Pazar_tf_id</code>: PAZAR database id.\n",
       "</p>\n",
       "<p>(3) Further criterias: \n",
       "</p>\n",
       "<p><code>min_ic</code> (minimum total information content of the matrix)\n",
       "</p>\n",
       "<p><code>length</code> (minimum sites length)\n",
       "</p>\n",
       "<p><code>sites</code> (minimum average sites number per base)\n",
       "</p>\n",
       "<p>When <code>all</code> is <code>TRUE</code>, it will get all the matrices \n",
       "and has higher priority over other options. \n",
       "Then ID has the second highest priority, \n",
       "and will ignore all the followiing options.\n",
       "The rest options are combined in search with AND, \n",
       "while multiple elements under one options have the logical operator OR.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>A <code>PFMatrixList</code> object.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Author(s)</h3>\n",
       "\n",
       "<p>Ge Tan\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "<p><code>getMatrixByID</code>,\n",
       "<code>getMatrixByName</code>\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre><code class='language-R'>  \n",
       "    library(JASPAR2014)\n",
       "    db &lt;- file.path(system.file(\"extdata\", package=\"JASPAR2014\"), \n",
       "                    \"JASPAR2014.sqlite\")\n",
       "    opts &lt;- list()\n",
       "    opts[[\"species\"]] &lt;- 9606\n",
       "    opts[[\"type\"]] &lt;- \"SELEX\"\n",
       "    opts[[\"all_versions\"]] &lt;- FALSE\n",
       "    siteList &lt;- getMatrixSet(db, opts)\n",
       "    siteList2 &lt;- getMatrixSet(JASPAR2014, opts)\n",
       "  \n",
       "</code></pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>TFBSTools</em> version 1.38.0 ]</div>\n",
       "</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{getMatrixSet}{Advanced JASPAR database search functions \\code{get\\_MatrixSet}}{getMatrixSet}\n",
       "\\aliasA{getMatrixSet,character-method}{getMatrixSet}{getMatrixSet,character.Rdash.method}\n",
       "\\aliasA{getMatrixSet,JASPAR2014-method}{getMatrixSet}{getMatrixSet,JASPAR2014.Rdash.method}\n",
       "\\aliasA{getMatrixSet,JASPAR2016-method}{getMatrixSet}{getMatrixSet,JASPAR2016.Rdash.method}\n",
       "\\aliasA{getMatrixSet,JASPAR2018-method}{getMatrixSet}{getMatrixSet,JASPAR2018.Rdash.method}\n",
       "\\aliasA{getMatrixSet,JASPAR2020-method}{getMatrixSet}{getMatrixSet,JASPAR2020.Rdash.method}\n",
       "\\aliasA{getMatrixSet,JASPAR2022-method}{getMatrixSet}{getMatrixSet,JASPAR2022.Rdash.method}\n",
       "\\aliasA{getMatrixSet,SQLiteConnection-method}{getMatrixSet}{getMatrixSet,SQLiteConnection.Rdash.method}\n",
       "%\n",
       "\\begin{Description}\n",
       "This function fetches matrix data for all matrices in the database \n",
       "matching criteria defined by the named arguments \n",
       "and returns a PFMatrixList object\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "  ## S4 method for signature 'character'\n",
       "getMatrixSet(x, opts)\n",
       "  ## S4 method for signature 'SQLiteConnection'\n",
       "getMatrixSet(x, opts)\n",
       "  ## S4 method for signature 'JASPAR2014'\n",
       "getMatrixSet(x, opts)\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{x}] \n",
       "a character vector of length 1 for the path of JASPAR SQLite file, \n",
       "a \\code{SQLiteConnection} object, or a \\code{JASPAR2014} object.\n",
       "\n",
       "\\item[\\code{opts}] \n",
       "a search options list. See more details below.\n",
       "\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\n",
       "The search options include three categories:\n",
       "\n",
       "(1) Database basic criterias: \n",
       "\n",
       "\\code{all=c(TRUE, FALSE)} \n",
       "\n",
       "\\code{ID}: a unique identifier for each model. \n",
       "CORE matrices always have a \"MAnnnnIDs.Version\".\n",
       "\n",
       "\\code{name}: The name of the transcription factor. \n",
       "As far as possible, the name is based on the standardized \n",
       "Entrez gene symbols. \n",
       "In the case the model describes a transcription factor hetero-dimer, \n",
       "two names are concatenated, such as RXR-VDR. \n",
       "In a few cases, different splice forms of the same gene \n",
       "have different binding specificity: in this case \n",
       "the splice form information is added to the name, \n",
       "based on the relevant literature.\n",
       "\n",
       "\\code{collection=c(\"CORE\", \"CNE\", \"PHYLOFACTS\", \"SPLICE\", \n",
       "    \"POLII\", \"FAM\", \"PBM\", \"PBM\\_HOMEO\", \"PBM\\_HLH\", \"UNVALIDATED\"}\n",
       "\n",
       "\\code{all\\_versions=c(FALSE,TRUE)}: We constantly update the profiles in\n",
       "JASPAR. Some profiles may have multiple versions. \n",
       "By default, only the latest version will be returned.\n",
       "\n",
       "\\code{species}: The species source for the sequences, in Latin (Homo sapiens)\n",
       "or NCBI tax IDs (9606).\n",
       "\n",
       "\\code{matrixtype=c(\"PFM\", \"PWM\", \"ICM\")}\n",
       "\n",
       "(2) Tags based criterias: \n",
       "\n",
       "\\code{class}: Structural class of the transcription factor, \n",
       "based on the TFCaT system. \n",
       "Examples: \"Zipper-Type\"\", \"Helix-Turn-Helix\", etc.\n",
       "\n",
       "\\code{type}: Methodology used for matrix construction: \n",
       "\"SELEX\", \"ChIP-seq\", \"PBM\", etc.\n",
       "\n",
       "\\code{tax\\_group}: Group of species, \n",
       "currently consisting of \"plants\", \"vertebrates\", \"insects\", \n",
       "\"urochordat\", \"nematodes\", \"fungi\".\n",
       "\n",
       "\\code{family}: Structural sub-class of the transcription factor, \n",
       "based on the TFCaT system.\n",
       "\n",
       "\\code{Acc}: A representative protein accession number in Genbank \n",
       "for the transcription factor. \n",
       "Human takes precedence if several exists. \n",
       "\n",
       "\\code{medline}: relevant publication reporting the sites \n",
       "used in the mode building.\n",
       "\n",
       "\\code{Pazar\\_tf\\_id}: PAZAR database id.\n",
       "\n",
       "(3) Further criterias: \n",
       "\n",
       "\\code{min\\_ic} (minimum total information content of the matrix)\n",
       "\n",
       "\\code{length} (minimum sites length)\n",
       "\n",
       "\\code{sites} (minimum average sites number per base)\n",
       "\n",
       "When \\code{all} is \\code{TRUE}, it will get all the matrices \n",
       "and has higher priority over other options. \n",
       "Then ID has the second highest priority, \n",
       "and will ignore all the followiing options.\n",
       "The rest options are combined in search with AND, \n",
       "while multiple elements under one options have the logical operator OR.\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "A \\code{\\LinkA{PFMatrixList}{PFMatrixList}} object.\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Author}\n",
       "Ge Tan\n",
       "\\end{Author}\n",
       "%\n",
       "\\begin{SeeAlso}\n",
       "\\code{\\LinkA{getMatrixByID}{getMatrixByID}},\n",
       "\\code{\\LinkA{getMatrixByName}{getMatrixByName}}\n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "  \n",
       "    library(JASPAR2014)\n",
       "    db <- file.path(system.file(\"extdata\", package=\"JASPAR2014\"), \n",
       "                    \"JASPAR2014.sqlite\")\n",
       "    opts <- list()\n",
       "    opts[[\"species\"]] <- 9606\n",
       "    opts[[\"type\"]] <- \"SELEX\"\n",
       "    opts[[\"all_versions\"]] <- FALSE\n",
       "    siteList <- getMatrixSet(db, opts)\n",
       "    siteList2 <- getMatrixSet(JASPAR2014, opts)\n",
       "  \n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "getMatrixSet             package:TFBSTools             R Documentation\n",
       "\n",
       "_\bA_\bd_\bv_\ba_\bn_\bc_\be_\bd _\bJ_\bA_\bS_\bP_\bA_\bR _\bd_\ba_\bt_\ba_\bb_\ba_\bs_\be _\bs_\be_\ba_\br_\bc_\bh _\bf_\bu_\bn_\bc_\bt_\bi_\bo_\bn_\bs '_\bg_\be_\bt__\bM_\ba_\bt_\br_\bi_\bx_\bS_\be_\bt'\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     This function fetches matrix data for all matrices in the database\n",
       "     matching criteria defined by the named arguments and returns a\n",
       "     PFMatrixList object\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     ## S4 method for signature 'character'\n",
       "     getMatrixSet(x, opts)\n",
       "       ## S4 method for signature 'SQLiteConnection'\n",
       "     getMatrixSet(x, opts)\n",
       "       ## S4 method for signature 'JASPAR2014'\n",
       "     getMatrixSet(x, opts)\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "       x: a character vector of length 1 for the path of JASPAR SQLite\n",
       "          file, a ‘SQLiteConnection’ object, or a ‘JASPAR2014’ object.\n",
       "\n",
       "    opts: a search options list. See more details below.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     The search options include three categories:\n",
       "\n",
       "     (1) Database basic criterias:\n",
       "\n",
       "     ‘all=c(TRUE, FALSE)’\n",
       "\n",
       "     ‘ID’: a unique identifier for each model.  CORE matrices always\n",
       "     have a \"MAnnnnIDs.Version\".\n",
       "\n",
       "     ‘name’: The name of the transcription factor.  As far as possible,\n",
       "     the name is based on the standardized Entrez gene symbols.  In the\n",
       "     case the model describes a transcription factor hetero-dimer, two\n",
       "     names are concatenated, such as RXR-VDR.  In a few cases,\n",
       "     different splice forms of the same gene have different binding\n",
       "     specificity: in this case the splice form information is added to\n",
       "     the name, based on the relevant literature.\n",
       "\n",
       "     ‘collection=c(\"CORE\", \"CNE\", \"PHYLOFACTS\", \"SPLICE\", \"POLII\",\n",
       "     \"FAM\", \"PBM\", \"PBM_HOMEO\", \"PBM_HLH\", \"UNVALIDATED\"’\n",
       "\n",
       "     ‘all_versions=c(FALSE,TRUE)’: We constantly update the profiles in\n",
       "     JASPAR. Some profiles may have multiple versions.  By default,\n",
       "     only the latest version will be returned.\n",
       "\n",
       "     ‘species’: The species source for the sequences, in Latin (Homo\n",
       "     sapiens) or NCBI tax IDs (9606).\n",
       "\n",
       "     ‘matrixtype=c(\"PFM\", \"PWM\", \"ICM\")’\n",
       "\n",
       "     (2) Tags based criterias:\n",
       "\n",
       "     ‘class’: Structural class of the transcription factor, based on\n",
       "     the TFCaT system.  Examples: \"Zipper-Type\"\", \"Helix-Turn-Helix\",\n",
       "     etc.\n",
       "\n",
       "     ‘type’: Methodology used for matrix construction: \"SELEX\",\n",
       "     \"ChIP-seq\", \"PBM\", etc.\n",
       "\n",
       "     ‘tax_group’: Group of species, currently consisting of \"plants\",\n",
       "     \"vertebrates\", \"insects\", \"urochordat\", \"nematodes\", \"fungi\".\n",
       "\n",
       "     ‘family’: Structural sub-class of the transcription factor, based\n",
       "     on the TFCaT system.\n",
       "\n",
       "     ‘Acc’: A representative protein accession number in Genbank for\n",
       "     the transcription factor.  Human takes precedence if several\n",
       "     exists.\n",
       "\n",
       "     ‘medline’: relevant publication reporting the sites used in the\n",
       "     mode building.\n",
       "\n",
       "     ‘Pazar_tf_id’: PAZAR database id.\n",
       "\n",
       "     (3) Further criterias:\n",
       "\n",
       "     ‘min_ic’ (minimum total information content of the matrix)\n",
       "\n",
       "     ‘length’ (minimum sites length)\n",
       "\n",
       "     ‘sites’ (minimum average sites number per base)\n",
       "\n",
       "     When ‘all’ is ‘TRUE’, it will get all the matrices and has higher\n",
       "     priority over other options.  Then ID has the second highest\n",
       "     priority, and will ignore all the followiing options.  The rest\n",
       "     options are combined in search with AND, while multiple elements\n",
       "     under one options have the logical operator OR.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     A ‘PFMatrixList’ object.\n",
       "\n",
       "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
       "\n",
       "     Ge Tan\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "     ‘getMatrixByID’, ‘getMatrixByName’\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "         library(JASPAR2014)\n",
       "         db <- file.path(system.file(\"extdata\", package=\"JASPAR2014\"), \n",
       "                         \"JASPAR2014.sqlite\")\n",
       "         opts <- list()\n",
       "         opts[[\"species\"]] <- 9606\n",
       "         opts[[\"type\"]] <- \"SELEX\"\n",
       "         opts[[\"all_versions\"]] <- FALSE\n",
       "         siteList <- getMatrixSet(db, opts)\n",
       "         siteList2 <- getMatrixSet(JASPAR2014, opts)\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?TFBSTools::getMatrixSet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "815830de-1fd0-47b9-9571-571dbc475bc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "PWM <- TFBSTools::getMatrixSet(\n",
    "    JASPAR2020,\n",
    "    opts = list(\n",
    "        collection = \"CORE\", \n",
    "        species = \"9606\", # taxid for Homo sapiens \n",
    "        all_versions = FALSE, \n",
    "        matrixtype = \"PWM\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a5cdf9ea-a0aa-4be1-b4ae-1e906341fec2",
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_names <- map_chr(\n",
    "    PWM@listData,\n",
    "    function(motif) {\n",
    "        paste0(motif@name, \"_\", motif@ID)\n",
    "    })\n",
    "names(PWM) <- motif_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f738ac6e-2960-4bea-9308-70c83b70833d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Adding coverage\n",
      "\n",
      "Adding peak set\n",
      "\n",
      "Searching For MACS2..\n",
      "\n",
      "Found with $path!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                            Group nCells nCellsUsed nReplicates\n",
      "EXP-00454-P1_PC02184-049 EXP-00454-P1_PC02184-049   4614        540           2\n",
      "EXP-00454-P1_PC02184-048 EXP-00454-P1_PC02184-048   2679        540           2\n",
      "EXP-00454-P1_PC02184-047 EXP-00454-P1_PC02184-047   3562        540           2\n",
      "EXP-00454-P1_PC02184-046 EXP-00454-P1_PC02184-046   1372        540           2\n",
      "EXP-00454-P1_PC02184-045 EXP-00454-P1_PC02184-045   7124        540           2\n",
      "EXP-00454-P1_PC02184-044 EXP-00454-P1_PC02184-044   5262        540           2\n",
      "EXP-00454-P1_PC02184-043 EXP-00454-P1_PC02184-043   4371        540           2\n",
      "EXP-00454-P1_PC02184-042 EXP-00454-P1_PC02184-042   2742        540           2\n",
      "EXP-00454-P1_PC02184-041 EXP-00454-P1_PC02184-041    981        540           2\n",
      "EXP-00454-P1_PC02184-040 EXP-00454-P1_PC02184-040    648        540           2\n",
      "EXP-00454-P1_PC02184-039 EXP-00454-P1_PC02184-039    685        540           2\n",
      "EXP-00454-P1_PC02184-038 EXP-00454-P1_PC02184-038   1583        540           2\n",
      "                         nMin nMax maxPeaks\n",
      "EXP-00454-P1_PC02184-049   40  500   150000\n",
      "EXP-00454-P1_PC02184-048   40  500   150000\n",
      "EXP-00454-P1_PC02184-047   40  500   150000\n",
      "EXP-00454-P1_PC02184-046   40  500   150000\n",
      "EXP-00454-P1_PC02184-045   40  500   150000\n",
      "EXP-00454-P1_PC02184-044   40  500   150000\n",
      "EXP-00454-P1_PC02184-043   40  500   150000\n",
      "EXP-00454-P1_PC02184-042   40  500   150000\n",
      "EXP-00454-P1_PC02184-041   40  500   150000\n",
      "EXP-00454-P1_PC02184-040   40  500   150000\n",
      "EXP-00454-P1_PC02184-039   40  500   150000\n",
      "EXP-00454-P1_PC02184-038   40  500   150000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Converged after 5 iterations!\n",
      "\n",
      "Adding peak matrix\n",
      "\n",
      "Adding Peak Annotations\n",
      "\n",
      "Saving ArchRProject...\n",
      "\n",
      "Loading ArchRProject...\n",
      "\n",
      "Successfully loaded ArchRProject!\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "Adding coverage\n",
      "\n",
      "Adding peak set\n",
      "\n",
      "Searching For MACS2..\n",
      "\n",
      "Found with $path!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                            Group nCells nCellsUsed nReplicates\n",
      "EXP-00454-P1_PC02184-049 EXP-00454-P1_PC02184-049   1481        540           2\n",
      "EXP-00454-P1_PC02184-048 EXP-00454-P1_PC02184-048    819        540           2\n",
      "EXP-00454-P1_PC02184-047 EXP-00454-P1_PC02184-047   1214        540           2\n",
      "EXP-00454-P1_PC02184-046 EXP-00454-P1_PC02184-046    479        479           2\n",
      "EXP-00454-P1_PC02184-045 EXP-00454-P1_PC02184-045   2097        540           2\n",
      "EXP-00454-P1_PC02184-044 EXP-00454-P1_PC02184-044   1812        540           2\n",
      "EXP-00454-P1_PC02184-043 EXP-00454-P1_PC02184-043   1419        540           2\n",
      "EXP-00454-P1_PC02184-042 EXP-00454-P1_PC02184-042    894        540           2\n",
      "EXP-00454-P1_PC02184-041 EXP-00454-P1_PC02184-041    401        401           2\n",
      "EXP-00454-P1_PC02184-040 EXP-00454-P1_PC02184-040    180        180           2\n",
      "EXP-00454-P1_PC02184-039 EXP-00454-P1_PC02184-039    657        540           2\n",
      "EXP-00454-P1_PC02184-038 EXP-00454-P1_PC02184-038    578        540           2\n",
      "                         nMin nMax maxPeaks\n",
      "EXP-00454-P1_PC02184-049   40  500   150000\n",
      "EXP-00454-P1_PC02184-048   40  500   150000\n",
      "EXP-00454-P1_PC02184-047   40  500   150000\n",
      "EXP-00454-P1_PC02184-046   40  439   150000\n",
      "EXP-00454-P1_PC02184-045   40  500   150000\n",
      "EXP-00454-P1_PC02184-044   40  500   150000\n",
      "EXP-00454-P1_PC02184-043   40  500   150000\n",
      "EXP-00454-P1_PC02184-042   40  500   150000\n",
      "EXP-00454-P1_PC02184-041   40  361   150000\n",
      "EXP-00454-P1_PC02184-040   40  140    90000\n",
      "EXP-00454-P1_PC02184-039   40  500   150000\n",
      "EXP-00454-P1_PC02184-038   40  500   150000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Converged after 6 iterations!\n",
      "\n",
      "Adding peak matrix\n",
      "\n",
      "Adding Peak Annotations\n",
      "\n",
      "Saving ArchRProject...\n",
      "\n",
      "Loading ArchRProject...\n",
      "\n",
      "Successfully loaded ArchRProject!\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "Adding coverage\n",
      "\n",
      "Adding peak set\n",
      "\n",
      "Searching For MACS2..\n",
      "\n",
      "Found with $path!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                            Group nCells nCellsUsed nReplicates\n",
      "EXP-00454-P1_PC02184-049 EXP-00454-P1_PC02184-049   6876        540           2\n",
      "EXP-00454-P1_PC02184-048 EXP-00454-P1_PC02184-048   4420        540           2\n",
      "EXP-00454-P1_PC02184-047 EXP-00454-P1_PC02184-047   5953        540           2\n",
      "EXP-00454-P1_PC02184-046 EXP-00454-P1_PC02184-046   2296        540           2\n",
      "EXP-00454-P1_PC02184-045 EXP-00454-P1_PC02184-045  10717        540           2\n",
      "EXP-00454-P1_PC02184-044 EXP-00454-P1_PC02184-044   9299        540           2\n",
      "EXP-00454-P1_PC02184-043 EXP-00454-P1_PC02184-043   6974        540           2\n",
      "EXP-00454-P1_PC02184-042 EXP-00454-P1_PC02184-042   4733        540           2\n",
      "EXP-00454-P1_PC02184-041 EXP-00454-P1_PC02184-041   1739        540           2\n",
      "EXP-00454-P1_PC02184-040 EXP-00454-P1_PC02184-040   1352        540           2\n",
      "EXP-00454-P1_PC02184-039 EXP-00454-P1_PC02184-039   2483        540           2\n",
      "EXP-00454-P1_PC02184-038 EXP-00454-P1_PC02184-038   3059        540           2\n",
      "                         nMin nMax maxPeaks\n",
      "EXP-00454-P1_PC02184-049   40  500   150000\n",
      "EXP-00454-P1_PC02184-048   40  500   150000\n",
      "EXP-00454-P1_PC02184-047   40  500   150000\n",
      "EXP-00454-P1_PC02184-046   40  500   150000\n",
      "EXP-00454-P1_PC02184-045   40  500   150000\n",
      "EXP-00454-P1_PC02184-044   40  500   150000\n",
      "EXP-00454-P1_PC02184-043   40  500   150000\n",
      "EXP-00454-P1_PC02184-042   40  500   150000\n",
      "EXP-00454-P1_PC02184-041   40  500   150000\n",
      "EXP-00454-P1_PC02184-040   40  500   150000\n",
      "EXP-00454-P1_PC02184-039   40  500   150000\n",
      "EXP-00454-P1_PC02184-038   40  500   150000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Converged after 5 iterations!\n",
      "\n",
      "Adding peak matrix\n",
      "\n",
      "Adding Peak Annotations\n",
      "\n",
      "Saving ArchRProject...\n",
      "\n",
      "Loading ArchRProject...\n",
      "\n",
      "Successfully loaded ArchRProject!\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "Adding coverage\n",
      "\n",
      "Adding peak set\n",
      "\n",
      "Searching For MACS2..\n",
      "\n",
      "Found with $path!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                            Group nCells nCellsUsed nReplicates\n",
      "EXP-00454-P1_PC02184-049 EXP-00454-P1_PC02184-049   1121        540           2\n",
      "EXP-00454-P1_PC02184-048 EXP-00454-P1_PC02184-048    567        540           2\n",
      "EXP-00454-P1_PC02184-047 EXP-00454-P1_PC02184-047    774        540           2\n",
      "EXP-00454-P1_PC02184-046 EXP-00454-P1_PC02184-046    275        275           2\n",
      "EXP-00454-P1_PC02184-045 EXP-00454-P1_PC02184-045   1271        540           2\n",
      "EXP-00454-P1_PC02184-044 EXP-00454-P1_PC02184-044   1022        540           2\n",
      "EXP-00454-P1_PC02184-043 EXP-00454-P1_PC02184-043    712        540           2\n",
      "EXP-00454-P1_PC02184-042 EXP-00454-P1_PC02184-042    516        516           2\n",
      "EXP-00454-P1_PC02184-041 EXP-00454-P1_PC02184-041    210        210           2\n",
      "EXP-00454-P1_PC02184-040 EXP-00454-P1_PC02184-040    126        126           2\n",
      "EXP-00454-P1_PC02184-039 EXP-00454-P1_PC02184-039     78         65           2\n",
      "EXP-00454-P1_PC02184-038 EXP-00454-P1_PC02184-038    319        319           2\n",
      "                         nMin nMax maxPeaks\n",
      "EXP-00454-P1_PC02184-049   40  500   150000\n",
      "EXP-00454-P1_PC02184-048   40  500   150000\n",
      "EXP-00454-P1_PC02184-047   40  500   150000\n",
      "EXP-00454-P1_PC02184-046   40  235   137500\n",
      "EXP-00454-P1_PC02184-045   40  500   150000\n",
      "EXP-00454-P1_PC02184-044   40  500   150000\n",
      "EXP-00454-P1_PC02184-043   40  500   150000\n",
      "EXP-00454-P1_PC02184-042   40  476   150000\n",
      "EXP-00454-P1_PC02184-041   40  170   105000\n",
      "EXP-00454-P1_PC02184-040   40   86    63000\n",
      "EXP-00454-P1_PC02184-039   40   40    32500\n",
      "EXP-00454-P1_PC02184-038   40  279   150000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Converged after 5 iterations!\n",
      "\n",
      "Adding peak matrix\n",
      "\n",
      "Adding Peak Annotations\n",
      "\n",
      "Saving ArchRProject...\n",
      "\n",
      "Loading ArchRProject...\n",
      "\n",
      "Successfully loaded ArchRProject!\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "Adding coverage\n",
      "\n",
      "Adding peak set\n",
      "\n",
      "Searching For MACS2..\n",
      "\n",
      "Found with $path!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                            Group nCells nCellsUsed nReplicates\n",
      "EXP-00454-P1_PC02184-049 EXP-00454-P1_PC02184-049   2261        540           2\n",
      "EXP-00454-P1_PC02184-048 EXP-00454-P1_PC02184-048    886        540           2\n",
      "EXP-00454-P1_PC02184-047 EXP-00454-P1_PC02184-047   1231        540           2\n",
      "EXP-00454-P1_PC02184-046 EXP-00454-P1_PC02184-046    460        460           2\n",
      "EXP-00454-P1_PC02184-045 EXP-00454-P1_PC02184-045   2472        540           2\n",
      "EXP-00454-P1_PC02184-044 EXP-00454-P1_PC02184-044   1793        540           2\n",
      "EXP-00454-P1_PC02184-043 EXP-00454-P1_PC02184-043   1371        540           2\n",
      "EXP-00454-P1_PC02184-042 EXP-00454-P1_PC02184-042    832        540           2\n",
      "EXP-00454-P1_PC02184-041 EXP-00454-P1_PC02184-041    328        328           2\n",
      "EXP-00454-P1_PC02184-040 EXP-00454-P1_PC02184-040    163        163           2\n",
      "EXP-00454-P1_PC02184-039 EXP-00454-P1_PC02184-039    200        200           2\n",
      "EXP-00454-P1_PC02184-038 EXP-00454-P1_PC02184-038    513        513           2\n",
      "                         nMin nMax maxPeaks\n",
      "EXP-00454-P1_PC02184-049   40  500   150000\n",
      "EXP-00454-P1_PC02184-048   40  500   150000\n",
      "EXP-00454-P1_PC02184-047   40  500   150000\n",
      "EXP-00454-P1_PC02184-046   40  420   150000\n",
      "EXP-00454-P1_PC02184-045   40  500   150000\n",
      "EXP-00454-P1_PC02184-044   40  500   150000\n",
      "EXP-00454-P1_PC02184-043   40  500   150000\n",
      "EXP-00454-P1_PC02184-042   40  500   150000\n",
      "EXP-00454-P1_PC02184-041   40  288   150000\n",
      "EXP-00454-P1_PC02184-040   40  123    81500\n",
      "EXP-00454-P1_PC02184-039   40  160   100000\n",
      "EXP-00454-P1_PC02184-038   40  473   150000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Converged after 5 iterations!\n",
      "\n",
      "Adding peak matrix\n",
      "\n",
      "Adding Peak Annotations\n",
      "\n",
      "Saving ArchRProject...\n",
      "\n",
      "Loading ArchRProject...\n",
      "\n",
      "Successfully loaded ArchRProject!\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "Adding coverage\n",
      "\n",
      "Adding peak set\n",
      "\n",
      "Searching For MACS2..\n",
      "\n",
      "Found with $path!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                            Group nCells nCellsUsed nReplicates\n",
      "EXP-00454-P1_PC02184-049 EXP-00454-P1_PC02184-049   2456        540           2\n",
      "EXP-00454-P1_PC02184-048 EXP-00454-P1_PC02184-048   1448        540           2\n",
      "EXP-00454-P1_PC02184-047 EXP-00454-P1_PC02184-047   1945        540           2\n",
      "EXP-00454-P1_PC02184-046 EXP-00454-P1_PC02184-046    856        540           2\n",
      "EXP-00454-P1_PC02184-045 EXP-00454-P1_PC02184-045   3672        540           2\n",
      "EXP-00454-P1_PC02184-044 EXP-00454-P1_PC02184-044   2989        540           2\n",
      "EXP-00454-P1_PC02184-043 EXP-00454-P1_PC02184-043   2292        540           2\n",
      "EXP-00454-P1_PC02184-042 EXP-00454-P1_PC02184-042   1786        540           2\n",
      "EXP-00454-P1_PC02184-041 EXP-00454-P1_PC02184-041    601        540           2\n",
      "EXP-00454-P1_PC02184-040 EXP-00454-P1_PC02184-040    451        451           2\n",
      "EXP-00454-P1_PC02184-039 EXP-00454-P1_PC02184-039    755        540           2\n",
      "EXP-00454-P1_PC02184-038 EXP-00454-P1_PC02184-038   1107        540           2\n",
      "                         nMin nMax maxPeaks\n",
      "EXP-00454-P1_PC02184-049   40  500   150000\n",
      "EXP-00454-P1_PC02184-048   40  500   150000\n",
      "EXP-00454-P1_PC02184-047   40  500   150000\n",
      "EXP-00454-P1_PC02184-046   40  500   150000\n",
      "EXP-00454-P1_PC02184-045   40  500   150000\n",
      "EXP-00454-P1_PC02184-044   40  500   150000\n",
      "EXP-00454-P1_PC02184-043   40  500   150000\n",
      "EXP-00454-P1_PC02184-042   40  500   150000\n",
      "EXP-00454-P1_PC02184-041   40  500   150000\n",
      "EXP-00454-P1_PC02184-040   40  411   150000\n",
      "EXP-00454-P1_PC02184-039   40  500   150000\n",
      "EXP-00454-P1_PC02184-038   40  500   150000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Converged after 5 iterations!\n",
      "\n",
      "Adding peak matrix\n",
      "\n",
      "Adding Peak Annotations\n",
      "\n",
      "Saving ArchRProject...\n",
      "\n",
      "Loading ArchRProject...\n",
      "\n",
      "Successfully loaded ArchRProject!\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "type_proj <- map(\n",
    "    type_proj,\n",
    "    function(proj) {\n",
    "        message(\"Adding coverage\")\n",
    "        proj <- addGroupCoverages(proj, groupBy = \"Sample\", force = TRUE)\n",
    "        message(\"Adding peak set\")\n",
    "        proj <- addReproduciblePeakSet(proj, groupBy = \"Sample\", force = TRUE)\n",
    "        message(\"Adding peak matrix\")\n",
    "        proj <- addPeakMatrix(proj, force = TRUE)\n",
    "        message(\"Adding Peak Annotations\")\n",
    "        proj <- addMotifAnnotations(\n",
    "            proj, \n",
    "            motifPWMs = PWM,\n",
    "            name = \"Motif\",\n",
    "            force = TRUE\n",
    "        )\n",
    "        proj <- saveArchRProject(proj)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01e8c91f-b2ce-417a-b34c-4ad7e091afa4",
   "metadata": {},
   "source": [
    "Output new annotations matrix for downstream use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "41ea2d5d-6436-4147-b979-f78b44725d69",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(\"output\"):\n",
      "“'output' already exists”\n"
     ]
    }
   ],
   "source": [
    "dir.create(\"output\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a16c8801-4626-4719-8c5d-874cd8dde141",
   "metadata": {},
   "outputs": [],
   "source": [
    "type_motif_mat <- map(\n",
    "    type_proj,\n",
    "    function(proj) {\n",
    "        anno <- getPeakAnnotation(proj)\n",
    "        matches <- readRDS(anno$Matches)\n",
    "        assays(matches)$matches\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4c4937ff-3818-4454-865d-5d973b50541d",
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_mat_output_files <- paste0(\n",
    "    \"output/peak-JASPAR-matches-\",\n",
    "    cell_types,\n",
    "    \"_\", Sys.Date(),\n",
    "    \".rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "82cd4cf7-b24d-41f4-a4cd-5bd6ac72c37c",
   "metadata": {},
   "outputs": [],
   "source": [
    "walk2(\n",
    "    type_motif_mat,\n",
    "    motif_mat_output_files,\n",
    "    saveRDS\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10c98748-9a8d-471e-ad77-f54b1c47bb30",
   "metadata": {},
   "source": [
    "### Store results in HISE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3da22f16-9541-42c3-9f85-87e58454af01",
   "metadata": {},
   "source": [
    "Update folder names for ArchR Projects and bundle the files as .tar for upload."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "b4efe021-6b3f-40fb-95f6-81496823bbbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "study_space_uuid <- \"40df6403-29f0-4b45-ab7d-f46d420c422e\"\n",
    "title <- paste(\"VRd TEA-seq T Cell JASPAR Annotations\", Sys.Date())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "257f2c26-a4ea-4720-a0c5-7b3a5a66d7b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_list <- as.list(\n",
    "    list.files(\n",
    "        \"output\", \n",
    "        pattern = \"peak-JASPAR-matches\",\n",
    "        full.names = TRUE\n",
    "))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "b2f89b0b-2147-43d1-b2f8-1b435ed721c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Authorization token invalid or expired.\"\n",
      "[1] \"Retrying...\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$files</dt>\n",
       "\t\t<dd><ol>\n",
       "\t<li>'output/peak-JASPAR-matches-t_cd4_cm_2023-10-09.rds'</li>\n",
       "\t<li>'output/peak-JASPAR-matches-t_cd4_em_2023-10-09.rds'</li>\n",
       "\t<li>'output/peak-JASPAR-matches-t_cd4_naive_2023-10-09.rds'</li>\n",
       "\t<li>'output/peak-JASPAR-matches-t_cd4_treg_2023-10-09.rds'</li>\n",
       "\t<li>'output/peak-JASPAR-matches-t_cd8_memory_2023-10-09.rds'</li>\n",
       "\t<li>'output/peak-JASPAR-matches-t_cd8_naive_2023-10-09.rds'</li>\n",
       "</ol>\n",
       "</dd>\n",
       "\t<dt>$traceId</dt>\n",
       "\t\t<dd>'cbdf162b-085d-4913-a95c-838b94b08c0c'</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$files] \\begin{enumerate}\n",
       "\\item 'output/peak-JASPAR-matches-t\\_cd4\\_cm\\_2023-10-09.rds'\n",
       "\\item 'output/peak-JASPAR-matches-t\\_cd4\\_em\\_2023-10-09.rds'\n",
       "\\item 'output/peak-JASPAR-matches-t\\_cd4\\_naive\\_2023-10-09.rds'\n",
       "\\item 'output/peak-JASPAR-matches-t\\_cd4\\_treg\\_2023-10-09.rds'\n",
       "\\item 'output/peak-JASPAR-matches-t\\_cd8\\_memory\\_2023-10-09.rds'\n",
       "\\item 'output/peak-JASPAR-matches-t\\_cd8\\_naive\\_2023-10-09.rds'\n",
       "\\end{enumerate}\n",
       "\n",
       "\\item[\\$traceId] 'cbdf162b-085d-4913-a95c-838b94b08c0c'\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$files\n",
       ":   1. 'output/peak-JASPAR-matches-t_cd4_cm_2023-10-09.rds'\n",
       "2. 'output/peak-JASPAR-matches-t_cd4_em_2023-10-09.rds'\n",
       "3. 'output/peak-JASPAR-matches-t_cd4_naive_2023-10-09.rds'\n",
       "4. 'output/peak-JASPAR-matches-t_cd4_treg_2023-10-09.rds'\n",
       "5. 'output/peak-JASPAR-matches-t_cd8_memory_2023-10-09.rds'\n",
       "6. 'output/peak-JASPAR-matches-t_cd8_naive_2023-10-09.rds'\n",
       "\n",
       "\n",
       "\n",
       "$traceId\n",
       ":   'cbdf162b-085d-4913-a95c-838b94b08c0c'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$files\n",
       "$files[[1]]\n",
       "[1] \"output/peak-JASPAR-matches-t_cd4_cm_2023-10-09.rds\"\n",
       "\n",
       "$files[[2]]\n",
       "[1] \"output/peak-JASPAR-matches-t_cd4_em_2023-10-09.rds\"\n",
       "\n",
       "$files[[3]]\n",
       "[1] \"output/peak-JASPAR-matches-t_cd4_naive_2023-10-09.rds\"\n",
       "\n",
       "$files[[4]]\n",
       "[1] \"output/peak-JASPAR-matches-t_cd4_treg_2023-10-09.rds\"\n",
       "\n",
       "$files[[5]]\n",
       "[1] \"output/peak-JASPAR-matches-t_cd8_memory_2023-10-09.rds\"\n",
       "\n",
       "$files[[6]]\n",
       "[1] \"output/peak-JASPAR-matches-t_cd8_naive_2023-10-09.rds\"\n",
       "\n",
       "\n",
       "$traceId\n",
       "[1] \"cbdf162b-085d-4913-a95c-838b94b08c0c\"\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "uploadFiles(\n",
    "    files = out_list,\n",
    "    studySpaceId = study_space_uuid,\n",
    "    title = title,\n",
    "    inputFileIds = atac_file_uuids,\n",
    "    store = \"project\",\n",
    "    doPrompt = FALSE\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f425145-cfb6-4e81-86a6-407de1ddfd93",
   "metadata": {},
   "source": [
    "# Session Info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e6aba05a-69c1-4b76-9ad0-abd1048c7dae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.1 (2023-06-16)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.6 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /opt/conda/lib/libopenblasp-r0.3.23.so;  LAPACK version 3.11.0\n",
       "\n",
       "Random number generation:\n",
       " RNG:     L'Ecuyer-CMRG \n",
       " Normal:  Inversion \n",
       " Sample:  Rejection \n",
       " \n",
       "locale:\n",
       " [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       \n",
       " [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   \n",
       " [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          \n",
       "[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   \n",
       "\n",
       "time zone: Etc/UTC\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] JASPAR2020_0.99.10                motifmatchr_1.22.0               \n",
       " [3] purrr_1.0.2                       BSgenome.Hsapiens.UCSC.hg38_1.4.5\n",
       " [5] BSgenome_1.68.0                   rtracklayer_1.60.1               \n",
       " [7] Biostrings_2.68.1                 XVector_0.40.0                   \n",
       " [9] rhdf5_2.44.0                      SummarizedExperiment_1.30.2      \n",
       "[11] Biobase_2.60.0                    MatrixGenerics_1.12.3            \n",
       "[13] Rcpp_1.0.11                       Matrix_1.6-1                     \n",
       "[15] GenomicRanges_1.52.0              GenomeInfoDb_1.36.2              \n",
       "[17] IRanges_2.34.1                    S4Vectors_0.38.1                 \n",
       "[19] BiocGenerics_0.46.0               matrixStats_1.0.0                \n",
       "[21] data.table_1.14.8                 stringr_1.5.0                    \n",
       "[23] plyr_1.8.8                        magrittr_2.0.3                   \n",
       "[25] ggplot2_3.4.3                     gtable_0.3.4                     \n",
       "[27] gtools_3.9.4                      gridExtra_2.3                    \n",
       "[29] ArchR_1.0.2                       dplyr_1.1.3                      \n",
       "[31] hise_2.16.0                      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] DBI_1.1.3                   bitops_1.0-7               \n",
       " [3] rlang_1.1.1                 compiler_4.3.1             \n",
       " [5] RSQLite_2.3.1               png_0.1-8                  \n",
       " [7] reshape2_1.4.4              vctrs_0.6.3                \n",
       " [9] pkgconfig_2.0.3             crayon_1.5.2               \n",
       "[11] fastmap_1.1.1               caTools_1.18.2             \n",
       "[13] utf8_1.2.3                  Rsamtools_2.16.0           \n",
       "[15] pracma_2.4.2                tzdb_0.4.0                 \n",
       "[17] DirichletMultinomial_1.42.0 bit_4.0.5                  \n",
       "[19] zlibbioc_1.46.0             cachem_1.0.8               \n",
       "[21] CNEr_1.36.0                 jsonlite_1.8.7             \n",
       "[23] blob_1.2.4                  rhdf5filters_1.12.1        \n",
       "[25] DelayedArray_0.26.7         uuid_1.1-1                 \n",
       "[27] Rhdf5lib_1.22.0             BiocParallel_1.34.2        \n",
       "[29] R6_2.5.1                    stringi_1.7.12             \n",
       "[31] IRkernel_1.3.2              assertthat_0.2.1           \n",
       "[33] R.utils_2.12.2              base64enc_0.1-3            \n",
       "[35] readr_2.1.4                 tidyselect_1.2.0           \n",
       "[37] abind_1.4-5                 yaml_2.3.7                 \n",
       "[39] codetools_0.2-19            curl_5.0.2                 \n",
       "[41] lattice_0.21-8              tibble_3.2.1               \n",
       "[43] KEGGREST_1.40.0             withr_2.5.0                \n",
       "[45] evaluate_0.21               BiocManager_1.30.22        \n",
       "[47] pillar_1.9.0                generics_0.1.3             \n",
       "[49] RCurl_1.98-1.12             IRdisplay_1.1              \n",
       "[51] hms_1.1.3                   munsell_0.5.0              \n",
       "[53] scales_1.2.1                xtable_1.8-4               \n",
       "[55] glue_1.6.2                  seqLogo_1.66.0             \n",
       "[57] tools_4.3.1                 TFMPvalue_0.0.9            \n",
       "[59] BiocIO_1.10.0               annotate_1.78.0            \n",
       "[61] GenomicAlignments_1.36.0    pbdZMQ_0.3-10              \n",
       "[63] XML_3.99-0.14               poweRlaw_0.70.6            \n",
       "[65] TFBSTools_1.38.0            AnnotationDbi_1.62.2       \n",
       "[67] colorspace_2.1-0            GenomeInfoDbData_1.2.10    \n",
       "[69] repr_1.1.6.9000             restfulr_0.0.15            \n",
       "[71] cli_3.6.1                   fansi_1.0.4                \n",
       "[73] S4Arrays_1.0.6              R.methodsS3_1.8.2          \n",
       "[75] digest_0.6.33               rjson_0.2.21               \n",
       "[77] farver_2.1.1                R.oo_1.25.0                \n",
       "[79] memoise_2.0.1               htmltools_0.5.6            \n",
       "[81] lifecycle_1.0.3             httr_1.4.7                 \n",
       "[83] GO.db_3.17.0                bit64_4.0.5                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb975889-3865-492c-a28b-d3d0f0a222eb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4",
   "language": "R",
   "name": "ir4"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
